### ISSUE TITLE
Python super() not implemented

### ISSUE THREAD

**Issue by shaolo1** [QQuick/Transcrypt#26]
The following code generates 
SyntaxError: super() is only valid in derived class constructors

```
        class foo:
            def bar(self):
                print('bar')

        class foo2(foo):
            def bar(self):
                print('bar2')
                super().bar()
                # foo.bar(self) # workaround


```

The error message seems to imply it works in derived constructors, but using it there generates the same error.

**Comment by JdeH** [comment]
That's right. Super was skipped in favor of the just explicitly mentioning the class name as in:

    class C (A, B):
        def __init__ (self, x, y):
            autoTester.check ('In C constructor')
            A.__init__ (self, x)
            B.__init__ (self, y)

This is one of the things that was left out to keep Transcrypt small.

**Comment by shaolo1** [comment]
I'll have to take a look at the translation, but super() should trivially convert the same as the explicit class name (ie, not be seen on the javascript side).

**Comment by JdeH** [comment]
I'll put it on my list of things to look into for the next version. If it's really that simple I'll probably add it.

**Comment by kovidgoyal** [comment]
You cannot just replace super() by a class identifier at compile time. That would break user code that defines a function named super(). Not to mention that the semantics of super() in python involve runtime evaluation of the MRO (method resolution order), this cannot be simulated at compile time in the general case, at least not without storing a ton of information about the MRO on every class. In fact, IMO, super() is an absolute anti-pattern in python. See for instance: https://fuhm.net/super-harmful/

**Comment by gbylenok** [comment]
I'd love to see this enhancement in a future version.

**Comment by AXGKl** [comment]
I would tend to not support it, more on kovidgoyal's side. Also regarding the readability advantage which Python has compared to e.g. JS:

```
class Button(Widget):
     def __init__(self, dom_mountpoint, props):
                  Widget.__init__(....)
                      
```
is more clear and explicit than whats happening than in super call.

**Comment by blubberdiblub** [comment]
I'd like to disagree with the negative opinions about `super()` here. It's not an anti-pattern at all and it promotes the DRY principle. I'm talking about Python 3 here, of course, as you still need to repeat the class name in Python 2.

Going with the example of @AXGKl that would look like this:

```python
class Button(Widget):
    def __init__(self, dom_mountpoint, props):
        super().__init__(....)
```

For me, there's no loss in readability here. It's calling the initialization function(s) of the superclass(es), whichever that may be. I don't need to be reminded which classes those are. It's doing everything it needs to do before I add the specialization. In a well-modularized code with well-separated functions, it should be possible to specify the base classes once and rely on that from that point forward and focus on more important items.

Of course you need to be somewhat disciplined when using `super()`, as it has some pitfalls if you aren't careful, but it's very well possible and easy enough to put it into good use. I'd like to contrast the aforementioned negative article about `super()` with a positive one: https://rhettinger.wordpress.com/2011/05/26/super-considered-super/

**Comment by AXGKl** [comment]
@blubberdiblub: 
Transcrypt has made huge progress in autumn towards 'completeness'. In July I was happy if there was ONE way to get stuff done. Meanwhile Jacques reached a state with the project where the transpiler does not need to force users into certain programming philosophies anymore, the user has the choice in most cases, just like in CPython. 

I was using super also all the time in CPython and only Transcrypt made me think about the difference of it versus nailing which super class to use in the call explicitely, making me do it also frequently now in CPython. Still its sometimes really useful to leave it dynamic, accoring to the current mro. Which can change, Jacques meanwhile implemented even `<Metaclass.>__new__`...

**=>** To get to a dynamic `super` we need `mro` as well.**

@JdeH : 

This would be a `mro` function, getting to it from the `__bases__` you already have:

```python
58 def mro(c, r=None):
 59     ''' assemble, we have only the bases'''
 60     l = 0
 61     if r is None:
 62         r = [c]
 63         l = 1
 64     for b in c.__bases__:
 65         if not b in r and not b == object:
 66             r.append(b)
 67             mro(b, r)
 68     if l:
 69         r.append(object)
 70     return r
```

**Comment by JdeH** [comment]
With reference to #258

**Comment by boxed** [comment]
This is pretty important I feel, but even more important is to fail in a nice way when hitting this. Now it transpiles without errors but Safari gives me an error on loading the .js file.

**Comment by JdeH** [comment]
We'll have super (), it's not that much extra core code.

**Comment by boxed** [comment]
Awesome.

**Comment by JdeH** [comment]
I've been playing with this and I think the C3 algorithm adds quite some overhead.
Signaling the use of super () on the other hand isn't hard, provided no attempt is made to take the context into account. 

The manual

https://docs.python.org/3/library/functions.html#super

says the following:

> There are two typical use cases for super. In a class hierarchy with single inheritance, super can be used to refer to parent classes without naming them explicitly, thus making the code more maintainable. This use closely parallels the use of super in other programming languages.
> 
> The second use case is to support cooperative multiple inheritance in a dynamic execution environment. This use case is unique to Python and is not found in statically compiled languages or languages that only support single inheritance. This makes it possible to implement “diamond diagrams” where multiple base classes implement the same method. Good design dictates that this method have the same calling signature in every case (because the order of calls is determined at runtime, because that order adapts to changes in the class hierarchy, and because that order can include sibling classes that are unknown prior to runtime).
> 

The manual refers doesn't the second case in detail, but refers to the article mentioned below under 'Opion 2'.

I've been reading quite a lot about super () and the C3 algorithm.

Seems there are two opinions on the value of `super ()`:

Opinion 1:

https://fuhm.net/super-harmful/

This mainly is about the unintuitiveness  and quirks of using  `super ()` in combination with complicated inheritance hierarchies, especially featuring diamond subgrahps. I agree on that point and think it's best to use explicit class names in such cases for clarity.

Especially having to use `super ()` in  base class to call a constructor in its sibling is completely unintuitive. The word 'super' refers to superclass, so it shouldn't be needed to call it in a baseclass to reach method on a sibling of that baseclass.


```
class A:
    def __init__ (self):
        print ('A')
        super () . __init__ ()  # Needed to call constructor of sibling B!!
        print ('a')
        
class B:
    def __init__ (self):
        print ('B')
        super () . __init__ ()
        print ('b')
        
class C (A, B):
    def __init__ (self):
        print ('C')
        super () .__init__ ()
        print ('c')

C ()
```


Opinion 2:

https://rhettinger.wordpress.com/2011/05/26/super-considered-super/

The argument named in favour of `super` are that if you don't name your base class explicitly, you can alter which class you inherit from without changing the code where you call the ancestors methods. The examples given to support this are simple inheritance hierarchies.

Also in this article is a quite extensive inventarisation of things to whatch out for in less simple cases. This is where a lot of caution is needed.



While I find that using `super` in non-diamond cases indeed honours the 'DRY' principle, I find that in complex cases it violates 'Explicit is better than implicit'.

I think C3 adss a lot of complexity to avoid problems with code that shouldn't have been written that way anyhow. Note that I think `super ()` is useful in simple and straight cases. But if it's needed to make a class hierarchy like the one above work, I consider it a design flaw.

So it's probably best to add super for the simple case, but put it under a switch. In the normal case encountering a call to super will lead to an error report. If the switch is on, the simple case, looking up one method in the ancestor hierarchy, will be supported. All other cases will require explicitly mentioning class names.

I believe this is a reasonable balance between costs and  benefits.

**Comment by kovidgoyal** [comment]
I'll just note that as far as the DRYness of super() is concerned, that is trivially achieved, by aliasing

```py
BaseClass = SomeClass

class Subclass(BaseClass):
   ...
```

Now you only need to rename SomeClass in one place. Not to mention that with modern refactoring tools, renaming a class everywhere it is used is trivial to do, except in the very rare case of a codebase with lots of difference classes all named with the exact same name. 

Also IMO even using super() in the simple inheritance case is an antipattern, because it makes your code much less readable. Now someone reading through your sub-class has to remember what the super class is everytime they see super(), and there is no easy way to jump to the definition of the superclass without first going to the top of the sub class to get the name of the super class. 

Basically, super() improves (marginally, over aliasing) maintainability for someone very familiar with the code, at the cost of readability for everyone else.

**Comment by boxed** [comment]
Maybe I missed something but I didn't understand why put it behind a switch. Supporting the single inheritance case seems like a no brainer and not doing it severely limits the usefulness of transcrypt as most OO code with even one inheritance should have a super().

**Comment by boxed** [comment]
Kovid can argue as much as he want, this isn't a debate on pythons design choices. Super is used in idiomatic Python code. That's a simple and obvious fact. If you are trying to make Python run in the browser you need to target Python, don't you?

Otherwise please don't call Transcrypt Python.

**Comment by JdeH** [comment]
As said it's started out as a Small Sane Subset, but now covers a large part of the language.
But it will never be as complete as PyPyJs.

**Comment by JdeH** [comment]
A restricted form of `super ()` has been committed.
It deals with most cases, even in a diamond hierarchy, while being fast and lean.
It is described in it's autotestlet:

http://sterlicht.alwaysdata.net/transcrypt.org/docs/html/supported_constructs.html#super

**Comment by JdeH** [comment]
As ever feel free to reopen if anything wrong.

**Comment by boxed** [comment]
As ever: amazing turn around time!
